domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java

```diff
@@ -497,6 +503,28 @@ private void init() throws Exception {
     }
 
     client = createPgPool(vertx, postgreSQLClientConfig);
+
+
+    vertx.setPeriodic(CHECK_FOR_QUERY_TIMEOUT_INTERVAL, PostgresClient::checkStaleRequests);
+
+  }
+
+  private static void checkStaleRequests(Long timerId) {
+    long now = System.currentTimeMillis();
+    for (SQLConnection conn : activeConnections) {
+      if (now >= conn.timeoutTime) {
+        conn.conn.cancelRequest(ar -> {
+          if (ar.succeeded()) {
+            log.warn(
+                String.format("Cancelling request due to timeout after : %d ms",
+                    now - conn.timeoutTime));
+          } else {
+            log.warn("Failed to send cancelling request", ar.cause());
+          }
+        });
+        activeConnections.remove(conn);
+      }
+    }
```

Comment:
The activeConnections CopyOnWriteArraySet is prone to race conditions.

Reply:
Assume two requests that have reached their timeoutTime. Calling cancelRequest for the first is an async call. Before returning the cancelRequest result other processing may happen, for example an SQL result for the second request is returned and processed, and the connection is reused to start a third SQL query. Then checkStaleRequests still has timeout information about the second request and may cancel the request on that connection, but it may cancel the third request.

Why don't you use one timer per request?

    long timerId = vertx.setTimer(queryTimeout, id -> {
      conn.conn.cancelRequest(...);
    });
    if request completes
      vertx.cancelTimer(timerId);

This does not require any considerations about race conditions.